### PR TITLE
[backport] Release connection between empty getMore responses 

### PR DESCRIPTION
**Pull request by vbabanin**
Backport of: https://github.com/mongodb/mongo-java-driver/pull/1925

> - Extract getMoreLoop to loop on empty batches while releasing the connection between each getMore, preventing pool starvation from idle tailable cursors (e.g., change streams)
> - Add test to verify connection is released and re-acquired between consecutive empty getMores
> 
> JAVA-6142

Note: 
- changes are adjusted to the current branch, as `AsyncCommandCursor` does not exists in `5.6.x`
- Additionally applied suggestion by @rozza from the main PR as well: https://github.com/mongodb/mongo-java-driver/pull/1925#pullrequestreview-4013208237.

- [x] Has AI automation been used? - Backport was done fully by Claude Code Sonnet 4.6.

### CODE REVIEW

File: driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java [hunk 1]
```diff
@@ -206,19 +206,22 @@ private void getMoreLoop(final AsyncConnection connection, final ServerCursor se
...
+                    callback.onResult(commandCursorResult.getResults(), null);
+        });
+    }
 
-                    getMoreLoop(connection, nextServerCursor, callback);
+    private void getMoreLoop(final ServerCursor localServerCursor, final SingleResultCallback<List<T>> funcCallback) {
+        getMore(localServerCursor, (nextBatch, t) -> {
+            if (t != null) {
+                funcCallback.onResult(null, t);
+            } else if (resourceManager.getServerCursor() == null || (nextBatch != null && !nextBatch.isEmpty())) {
```
Comment (nhachicha) [inline driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java:217]: Interesting, `claude-review` flagged this as a potential _double volatile read_ and claimed that a concurrent `close()` could cause  `assertNotNull` to fail on the second read. However when I pushed back on this it realises the following 

```
You're right to push back — that claim was wrong.                                                                                                                                                                
                                                                                                                                     
  Looking at the actual state machine:                                                                                                                                                                             
                                                                                                                                                                                                                   
  1. When next() starts, tryStartOperation() transitions state from IDLE → OPERATION_IN_PROGRESS (line 148)                                                                                                        
  2. If close() is called concurrently while in OPERATION_IN_PROGRESS, it transitions to CLOSE_PENDING (line 186) — but it does not call doClose() and does not call unsetServerCursor()                           
  3. doClose() only runs later, when endOperation() sees CLOSE_PENDING (line 166-168)                                                                                                                              
  4. endOperation() is called after the next() callback completes                                                                                                                                                  
                                                                                                                                                                                                                   
  So during the entire execution of getMoreLoop's callback (lines 214-224), the state is OPERATION_IN_PROGRESS or CLOSE_PENDING. In neither case does serverCursor get unset — that only happens inside doClose(), 
  which is deferred until endOperation().                                                                                                                                                                          
                                                                                                                                                                                                                   
  This means the double volatile read on lines 217/223 is actually safe. The serverCursor cannot change between the two reads because:                                                                             
  - setServerCursor() asserts state.inProgress() and is only called from within the operation itself (line 208)
  - unsetServerCursor() is only called from doClose(), which cannot run until endOperation()                                                                                                                       
  - No other code path modifies serverCursor during an in-progress operation                                                         
```

Comment (nhachicha) [reply driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java:217]: Note:
This is still a warning reported by the IDE 

Image

File: driver-core/src/test/functional/com/mongodb/internal/operation/AsyncCommandBatchCursorTest.java [hunk 2]
```diff
@@ -252,6 +256,32 @@ void closeShouldNotResetOriginalTimeout(final boolean disableTimeoutResetWhenClo
...
+            cb.onResult(new BsonDocument("cursor",
+                    new BsonDocument("ns", new BsonString(NAMESPACE.getFullName()))
+                            .append("id", new BsonInt64(callCount.incrementAndGet() < 3 ? 1 : 0))
+                            .append("nextBatch", new BsonArrayWrapper<>(new BsonArray()))), null);
+            return null;
+        }).when(mockConnection).commandAsync(eq(NAMESPACE.getDatabaseName()),
+                argThat(doc -> doc.containsKey("getMore")), any(), any(), any(), any(), any());
+
+        when(serverDescription.getType()).thenReturn(ServerType.STANDALONE);
+        createBatchCursor(0).next((result, t) -> {
```
Comment (nhachicha) [inline driver-core/src/test/functional/com/mongodb/internal/operation/AsyncCommandBatchCursorTest.java]: [nit]
wrap with `try-with-resource`
```Java
try (AsyncCommandBatchCursor<Document> commandBatchCursor = createBatchCursor(0)) {
            commandBatchCursor.next((result, t) -> {
                assertNotNull(result);
                assertTrue(result.isEmpty());
                assertNull(t);
            });
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
```

**Review by nhachicha — DISMISSED** [review]

File: driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java (hunk 1 above)
Comment (vbabanin) [reply driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java:217]: > However when I pushed back on this it realises the following

I agree with your analysis; that the code is functionally safe due to the state machine invariants. The `AsyncCommandCursor` and `CommandCursor` are not allowed to be used concurrently and protected by the state machine. The IDE warning is still legitimate from a static analysis perspective - it can't infer these runtime invariants. Therefore, it is safe to leave it as it is.

File: driver-core/src/test/functional/com/mongodb/internal/operation/AsyncCommandBatchCursorTest.java (hunk 2 above)
Comment (vbabanin) [reply driver-core/src/test/functional/com/mongodb/internal/operation/AsyncCommandBatchCursorTest.java]: Good catch. I just noticed that we do not do this consistently in all places. So I added `try-with-resource` everywhere in this test class - and removed catch as it is redundant, exception will fail the test anyway, wrapping was unnecessary.

File: driver-core/src/test/functional/com/mongodb/internal/operation/AsyncCommandBatchCursorTest.java [hunk 3]
```diff
@@ -207,8 +202,6 @@ void closeShouldResetTimeoutContextToDefaultMaxTime() {
         try (AsyncCommandBatchCursor<Document> commandBatchCursor = createBatchCursor(maxTimeMS)) {
             // verify that the `maxTimeMS` override was applied
             timeoutContext.runMaxTimeMS(remainingMillis -> assertTrue(remainingMillis <= maxTimeMS));
-        } catch (Exception e) {
```
Comment (nhachicha) [inline driver-core/src/test/functional/com/mongodb/internal/operation/AsyncCommandBatchCursorTest.java:210]: Why was this wrapped in RuntimeException in the first place?

**Review by nhachicha — APPROVED** [review]
Thanks for updating the try-with-resources.
Will you backport https://github.com/mongodb/mongo-java-driver/pull/1928/commits/52898a57c00cca3de3d0d2518273865005a7efb3 to `main`?

**Comment by vbabanin** [comment]
> Will you backport https://github.com/mongodb/mongo-java-driver/commit/52898a57c00cca3de3d0d2518273865005a7efb3 to main?

i will create another cleanup PR for main.
